### PR TITLE
Add QNX support

### PR DESCRIPTION
**Pull request by jbreitbart**
It is not very well tested, but seems to get the job done.

### CODE REVIEW

File: src/libdistgen.cpp [hunk 1]
```diff
@@ -175,7 +191,34 @@ double distgend_is_membound_scaled(distgend_configT config) {
 static void *thread_benchmark(void *arg) {
 	thread_argsT *thread_args = (thread_argsT *)arg;
 	double *ret = (double *)calloc(1, sizeof(double));
+#ifdef __QNXNTO__
```
Comment (spickartz) [inline src/libdistgen.cpp]: Why don't you move this into a separate function `set_affinity` and call it for both the linux and the QNX variant at this place? We shouldn't run into NUMA effects calling it here, should we?

File: src/libdistgen.cpp [hunk 2]
```diff
@@ -175,7 +191,34 @@ double distgend_is_membound_scaled(distgend_configT config) {
...
+	// QNX does not support pthread thread affinity. Set affinity here.
+	// TODO duplicated code in distgen_internal.cpp
+	unsigned int number_of_cpus = RMSK_SIZE(_syspage_ptr->num_cpu);
+	const int size = number_of_cpus * sizeof(unsigned int) * 2;
+	
+	unsigned int *rsizep = (unsigned int*)malloc(size);
+	if (rsizep == nullptr) {
```
Comment (spickartz) [inline src/libdistgen.cpp]: Why not `assert(rsizep != nullptr)`?

File: src/libdistgen.cpp [hunk 3]
```diff
@@ -175,7 +191,34 @@ double distgend_is_membound_scaled(distgend_configT config) {
...
+	the thread can run on. */
+	unsigned int *rmaskp = rsizep + 1;
+	RMSK_SET(thread_args->tid, rmaskp);
+
+	/* Set the inherit mask. Call this macro once for each
+	processor the thread's children can run on. */
+	unsigned int *imaskp = rmaskp + number_of_cpus;
+	RMSK_SET(thread_args->tid, imaskp);
+
+	if ( ThreadCtl( _NTO_TCTL_RUNMASK_GET_AND_SET_INHERIT, rsizep) == -1) {
```
Comment (spickartz) [inline src/libdistgen.cpp]: See comment above.

File: src/libdistgen.cpp [hunk 4]
```diff
@@ -190,7 +233,9 @@ static void *thread_benchmark(void *arg) {
 			*ret += temp / (t2 - t1);
 		}
 	}
-
+#ifdef __QNXNTO__
+	free (rsizep);
```
Comment (spickartz) [inline src/libdistgen.cpp]: Okay, I see - we would need some kind of cleanup function to use the approach I suggest above. I assume `rsizep` is required for the whole lifetime of the thread.

File: src/libdistgen.cpp (hunk 3 above)
Comment (jbreitbart) [reply src/libdistgen.cpp]: Because this would remove the function call if compiled without asserts.

**Comment by jbreitbart** [comment]
I am going to merge this as is. I agree that in general there should be a function to make the QNX parts at the beginning of `thread_benchmark` ... but then again, I am not sure if it will be used much and we may just remove it later again :smile:.

**Comment by weidendo** [comment]
Is this something to add also the the distgen benchmark itself?

**Comment by jbreitbart** [comment]
Is distgen still using OpenMP? I am not sure if that is even supported on QNX. So no, it probably makes no sense.

**Comment by weidendo** [comment]
Does it help portability to use pthreads? I have no idea about OpenMP support in the embedded space.

**Comment by jbreitbart** [comment]
pthreads is certainly better. If you are using Linux you can probably also get OpenMP support, but if you are not than you have a hard time.

**Pull request merged**